### PR TITLE
Llama3-Vision feature-complete (batch 32 / paged attention / tracing)

### PR DESCRIPTION
**Pull request by cglagovichTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/15058)

### What's changed
This PR enables features for the Llama-Vision model:
- Batch > 1 support
- Tracing
- Paged attention
- New RoPE

In addition, it refactors the generation code for Llama Text and Llama Vision. This is to help vLLM integration for all models in this shared impl. 

### Checklist
- [x] T3K pipelines https://github.com/tenstorrent/tt-metal/actions/runs/12166448194
- [ ] All Post Commit https://github.com/tenstorrent/tt-metal/actions/runs/12182434194

### CODE REVIEW

**Review by tt-rkim — APPROVED** [review]

File: models/demos/llama3/tt/generator.py [hunk 1]
```diff
@@ -368,14 +552,14 @@ def sample(logits):
         )
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator.py:552]: This line `logits = logits.view(1, 1, self.model_args.vocab_size)` should be added in `generate()` after prefill_forward_single_user and before sample(logits) now that process_output_prefill returns 1d output

File: models/perf/benchmarking_utils.py [hunk 2]
```diff
@@ -16,6 +16,24 @@ def __init__(self):
...
+            self.iteration = iteration
+
+        def __enter__(self):
+            self.profiler.start(self.step_name, self.iteration)
+            return self.profiler
+
+        def __exit__(self, exc_type, exc_val, exc_tb):
+            self.profiler.end(self.step_name, self.iteration)
+            return False
+
```
Comment (skhorasganiTT) [inline models/perf/benchmarking_utils.py:36]: Nice

**Review by mtairum — APPROVED** [review]
Sent you my review comments through slack.

After you address them it's good to merge, from llama3-text side of things.
Also make sure the custom test pipelines are passing!

File: models/demos/llama3/demo/tiny_demo.py [hunk 3]
```diff
@@ -0,0 +1,590 @@
...
+                context_text = ""
+        except Exception as e:
+            logger.error(f"Error fetching context from URL: {context_url}. Error: {str(e)}")
+            context_text = ""
+
+    return context_text
+
+
+# load from json, return as a list
+def load_inputs(user_input, batch):
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/demos/llama3/demo/tiny_demo.py:64]: demo.py will continue to exist until the tiny demo with generator and vLLM integration is fully ready to superseed it.
So we definitely should consolidate it. We can also open a new issue and revisit the tiny demo so this PR is not blocked.

**Comment by cglagovichTT** [comment]
TODO: Update README with vision info

**Comment by cglagovichTT** [comment]
Kicking T3K CI after addressing comments https://github.com/tenstorrent/tt-metal/actions/runs/12187056001

File: models/demos/llama3/demo/simple_vision_demo.py [hunk 4]
```diff
@@ -42,7 +43,22 @@ def sample(logits):
...
+            next_token = torch.argmax(logits[:, -1], dim=-1)
+
+        next_tokens = next_token.reshape(-1)
+        texts = [tokenizer.decode([next_tokens[i].item()]) for i in range(len(next_tokens))]
+        return next_tokens, texts
+
+    return sample
+
+
+def create_multimodal_model(mesh_device, max_batch_size, max_seq_len, dtype=ttnn.bfloat16, use_paged_kv_cache=False):
```
[the remark this answers is not in the document]
Comment (cglagovichTT) [reply models/demos/llama3/demo/simple_vision_demo.py:44]: I removed `test_llama_vision_model` because it is broken and not useful right now -- we will use real benchmarks to get accuracy info for llama vision.

**Comment by cglagovichTT** [comment]
Thank you for the comments @skhorasganiTT and @mtairum !

**Review by skhorasganiTT — APPROVED** [review]

**Comment by skhorasganiTT** [comment]
@cglagovichTT One more thing to note (doesn't need to block this PR), the n300 demo should be moved from the t3k demo pipeline to the single-card demo pipeline (in the n300 perf demo sub-job here: https://github.com/tenstorrent/tt-metal/blob/main/tests/scripts/single_card/run_single_card_demo_tests.sh#L82)

**Review by yieldthought — APPROVED** [review]
🔥

**Comment by cglagovichTT** [comment]
Kicking Demo T3K CI after fixing copy_host_to_device bug I introduced https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by mtairum — APPROVED** [review]
